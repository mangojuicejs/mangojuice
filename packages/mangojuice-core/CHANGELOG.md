# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.3"></a>
## 0.1.3 (2017-07-12)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))
* **core:** recursive nesting works now ([c462be5](https://github.com/mangojuicejs/mangojuice/commit/c462be5))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)
