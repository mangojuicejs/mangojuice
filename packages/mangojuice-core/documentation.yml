toc:
  - name: Basic
    description: |
      The most common APIs that you will work with most of the time
    children:
      - Block
      - LogicBase
      - cmd
      - run
      - mount
      - task
      - cancel
      - child
      - depends
  - name: Extras
    description: |
      Other pieces of MangoJuice
    children:
      - Process
      - Command
      - TaskMeta
      - DefaultLogger
      - ComputedField
      - Mounter
      - defineCommand
      - ensureCommand
      - decorateLogic
      - handleBefore
      - handleAfter
      - handle
      - call
      - logicOf
      - observe
      - procOf
      - bind
      - hydrate
      - delay
