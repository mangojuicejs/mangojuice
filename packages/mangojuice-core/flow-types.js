// @flow
type CommandType<M> = {
  is: (...args: any) => bool,
  model: M,
};

type TaskObject = {
  success: (cmd: any) => TaskObject,
  fail: (cmd: any) => TaskObject,
  args: (...args: Array<any>) => TaskObject,
  engine: (exec: any) => TaskObject
};

type DependsType = {
  compute: (Function) => DependsType
};

type LogicConfig<MetaType> = {
  initCommands?: any,
  meta?: MetaType
};

type ConfiguredLogic = {
};

type BlockType = {
  createModel: (...args: any) => Object,
  Logic: Class<*>,
  View?: any
};

declare module "mangojuice-core" {
  declare export function cmd(): any;

  declare export function debounce(): any;

  declare export function throttle(): any;

  declare export function task(fn: Function): TaskObject;

  declare export function delay(ms: number): Promise<void>;

  declare export function cancel(cmd: any): any;

  declare export function logicOf<T>(model: Object, clazz?: Class<T>): T;

  declare export function child<T>(logicClass: Class<{ +config?: (...args: T) => any }>, ...args: T): ConfiguredLogic;

  declare export function depends(...args: any): DependsType;

  declare export function observe(model: Object, destroyeded: Promise<any>, handler: Function): void;

  declare export function run(block: BlockType, opts?: Object): any;

  declare export function bind(block: BlockType, opts?: Object): any;

  declare export function hydrate(block: BlockType, model: Object): any;

  declare export var utils: any;

  declare export class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
    +model: ModelType;
    +shared: SharedType;
    +meta: MetaType;

    hub(cmd: CommandType<*>): ?Promise<any>;

    hubBefore(cmd: CommandType<*>): ?Promise<any>;

    hubAfter(cmd: CommandType<*>): ?Promise<any>;

    port(exec: (cmd: any) => Promise<any>, destroyed: Promise<any>): ?Promise<any>;

    children(): {[k: $Enum<ModelType>]: Class<*> | ConfiguredLogic};

    computed(): {[k: $Enum<ModelType>]: (() => any) | DependsType};

    config(...args: any): LogicConfig<MetaType>;
  }

  declare export default {|
    +LogicBase: typeof LogicBase,
    +cmd: typeof cmd,
    +debounce: typeof debounce,
    +throttle: typeof throttle,
    +task: typeof task,
    +delay: typeof delay,
    +cancel: typeof cancel,
    +logicOf: typeof logicOf,
    +child: typeof child,
    +depends: typeof depends,
    +observe: typeof observe,
    +run: typeof run,
    +bind: typeof bind,
    +hydrate: typeof hydrate,
    +utils: typeof utils
  |};
}
