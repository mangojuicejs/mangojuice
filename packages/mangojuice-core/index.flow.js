// @flow

type NestType<A, T, Y, U> = {
  +logic: A,
  args: (c1?: T, c2?: Y, c3?: U) => NestType<A, T, Y, U>,
  handler: () => NestType<A, T, Y, U>,
  singleton: (?Boolean) => NestType<A, T, Y, U>
};

type DependsType = {
  compute: (Function) => DependsType
};

interface ConfigFunction<MetaType, T, Y, U> {
  +config?: (c1?: T, c2?: Y, c3?: U) => ConfigType<MetaType>;
};

type ConfigType<M> = {
  initCommands?: any,
  meta?: M
};

declare class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
  +model: ModelType;
  +shared: SharedType;
  +meta: MetaType;
  +destroy: Promise<any>;

  port(): ?Promise<any>;

  children(): {[key: $Enum<ModelType>]: NestType<*, *, *, *>};

  computed(): {[k: $Enum<ModelType>]: () => any | DependsType};

  exec(): Promise<any>;

  depends(...deps: Array<Object>): DependsType;

  nest<T, Y, U, L: ConfigFunction<MetaType, T, Y, U>&LogicBase<*,SharedType,*>>(logic: L): NestType<L, T, Y, U>;
}
