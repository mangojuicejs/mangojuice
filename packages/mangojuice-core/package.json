{
  "name": "mangojuice-core",
  "version": "0.1.6",
  "description": "",
  "author": {
    "name": "Artem Artemev",
    "email": "viiv.c58@gmail.com"
  },
  "bugs": {
    "url": "https://github.com/mangojuicejs/mangojuice/issues"
  },
  "homepage": "https://github.com/mangojuicejs/mangojuice#readme",
  "repository": "https://github.com/mangojuicejs/mangojuice",
  "license": "MIT",
  "main": "index.js",
  "rollup": {
    "moduleName": "Mjs"
  }
}
