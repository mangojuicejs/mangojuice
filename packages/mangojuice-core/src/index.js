import Process from './classes/Process';
import Command from './classes/Command';
import LogicBase from './classes/LogicBase';
import DefaultLogger from './classes/DefaultLogger';
import cmd from './core/cmd/cmd';
import cancel from './core/cmd/cancel';
import defineCommand from './core/cmd/defineCommand';
import ensureCommand from './core/cmd/ensureCommand';
import child from './core/logic/child';
import depends from './core/logic/depends';
import observe from './core/logic/observe';
import { handle, handleAfter, handleBefore } from './core/logic/handle';
import logicOf from './core/logic/logicOf';
import procOf from './core/logic/procOf';
import decorateLogic from './core/logic/decorateLogic';
import delay from './core/task/delay';
import task from './core/task/task';
import callTask from './core/task/callTask';
import { CANCEL } from './classes/AsyncTask';
import run from './core/run/run';
import bind from './core/run/bind';
import mount from './core/run/mount';
import hydrate from './core/run/hydrate';
import * as utils from './core/utils';
import * as config from './config';


export {
  // Constants
  CANCEL,
  // Classes
  Process,
  Command,
  LogicBase,
  DefaultLogger,
  // Core
  cmd,
  cancel,
  defineCommand,
  ensureCommand,
  child,
  depends,
  observe,
  handle,
  handleAfter,
  handleBefore,
  logicOf,
  procOf,
  decorateLogic,
  delay,
  task,
  callTask,
  run,
  bind,
  mount,
  hydrate,
  utils,
  config
};
