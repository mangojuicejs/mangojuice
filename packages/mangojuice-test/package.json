{
  "name": "mangojuice-test",
  "version": "2.0.0-y.0",
  "description": "",
  "author": {
    "name": "Artem Artemev",
    "email": "viiv.c58@gmail.com"
  },
  "bugs": {
    "url": "https://github.com/mangojuicejs/mangojuice/issues"
  },
  "homepage": "https://github.com/mangojuicejs/mangojuice#readme",
  "repository": "https://github.com/mangojuicejs/mangojuice",
  "license": "MIT",
  "main": "index.js",
  "rollup": {
    "moduleName": "Mjs.Test",
    "bundledDependencies": [],
    "moduleGlobals": {
      "mangojuice": "Mjs"
    }
  },
  "dependencies": {
    "mangojuice-core": "2.0.0-y.0"
  }
}
