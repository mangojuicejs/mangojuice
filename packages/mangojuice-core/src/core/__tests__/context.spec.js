import { child, context, logicOf, procOf, message } from 'mangojuice-core';
import { runWithTracking } from 'mangojuice-test';


describe('child', () => {
  describe('#create', () => {
    it('should create context and available in child', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c)
      });
      class ChildLogic {
        create() {
          return context(TEST_CONTEXT).get((ctx) => ({
            fromContext: ctx.test.ctx,
            argsFromContext: ctx.test.args
          }))
        }
      }
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { a: child(ChildLogic).create(1,2,3) }
          ];
        }
      }

      const { app } = runWithTracking({ app: { Logic: Test } });

      expect(app.model).toMatchSnapshot();
    });

    it('should be able to create context after child created', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c)
      });
      class ChildLogic {
        create() {
          return context(TEST_CONTEXT).get((ctx) => ({}))
        }
        update() {
          return context(TEST_CONTEXT).get((ctx) => ({
            fromContext: ctx.test.ctx,
            argsFromContext: ctx.test.args
          }))
        }
      }
      class Test {
        create() {
          return [
            { a: child(ChildLogic).create(1,2,3) },
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { a: child(ChildLogic).update(() => ({})) },
          ];
        }
      }

      const { app, errors } = runWithTracking({ app: { Logic: Test }, expectErrors: true });

      expect(app.model).toMatchSnapshot();
      expect(errors).toMatchSnapshot();
    })

    it('should destroy context with logic which created a context', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return { ...evt } }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
        anotherTest: child(ContextChildLogic).create(b, c, a)
      });
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) }
          ];
        }
      }

      const { app } = runWithTracking({ app: { Logic: Test } });
      const processes = [
        procOf(app.model),
        procOf(app.model.ctx),
        procOf(app.model.ctx.test),
        procOf(app.model.ctx.anotherTest)
      ];
      processes[0].destroy();

      expect(app.model).toMatchSnapshot();
      expect(processes.map(x => x.destroyed)).toMatchSnapshot();
    });
    it('should be able to create same context in different subtrees', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return { ...evt } }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
        anotherTest: child(ContextChildLogic).create(b, c, a)
      });
      class ChildLogic {
        create(...args) {
          return { ctx: context(TEST_CONTEXT).create(...args) }
        }
      }
      class Test {
        create() {
          return {
            a: child(ChildLogic).create(1,2,3),
            b: child(ChildLogic).create(3,2,1)
          };
        }
      }

      const { app } = runWithTracking({ app: { Logic: Test } });

      expect(app.model).toMatchSnapshot();
    });
    it('should throw an error if context already created - in root', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return { ...evt } }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
        anotherTest: child(ContextChildLogic).create(b, c, a)
      });
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { anotherCtx: context(TEST_CONTEXT).create(3,2,1) }
          ];
        }
      }

      const { app, errors } = runWithTracking({ app: { Logic: Test }, expectErrors: true });

      expect(app.model).toMatchSnapshot();
      expect(errors).toMatchSnapshot();
    })
    it('should override a parent context if same created in child', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return { ...evt } }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
        anotherTest: child(ContextChildLogic).create(b, c, a)
      });
      class ChildLogic {
        create(...args) {
          return [
            { ctx: context(TEST_CONTEXT).create(...args) },
            { child: child(AnotherChildLogic).create() }
          ]
        }
        updateFromContext() {
          return context(TEST_CONTEXT).get((ctx) => {
            fromContext: ctx.test.args
          })
        }
      }
      class AnotherChildLogic {
        create(...args) {
          return context(TEST_CONTEXT).get((ctx) => ({
            fromContext: ctx.test.args
          }))
        }
      }
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { child: child(ChildLogic).create(2,2,2) },
            { anotherChild: child(AnotherChildLogic).create(1,1,1) },
          ];
        }
      }

      const { app, errors } = runWithTracking({ app: { Logic: Test } });
      procOf(app.model.child).exec(logicOf(app.model.child).updateFromContext);

      expect(app.model).toMatchSnapshot();
      expect(errors).toMatchSnapshot();
    });
    it('should be able to create multiple contexts', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return { ...evt } }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
      });
      const ANOTHER_CONTEXT = (a, b, c) => ({
        anotherTest: child(ContextChildLogic).create(b, c, a)
      });
      class ChildLogic {
        create(...args) {
          return [
            context(TEST_CONTEXT).get((ctx) => ({
              fromTestContext: ctx.test.args
            })),
            context(ANOTHER_CONTEXT).get((ctx) => ({
              fromAnotherContext: ctx.anotherTest.args
            }))
          ]
        }
      }
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { anotherCtx: context(ANOTHER_CONTEXT).create(3,2,1) },
            { child: child(ChildLogic).create() }
          ];
        }
      }

      const { app, errors } = runWithTracking({ app: { Logic: Test } });

      expect(app.model).toMatchSnapshot();
    });
  });

  describe('#update', () => {
    it('should send update context from root', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return { ...evt } }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
        anotherTest: child(ContextChildLogic).create(b, c, a)
      });
      class ChildLogic {
        create() {
          return context(TEST_CONTEXT).get((ctx) => ({
            fromContext: ctx.test.ctx,
            argsFromContext: ctx.test.args
          }))
        }
      }
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { a: child(ChildLogic).create(1,2,3) },
            context(TEST_CONTEXT).update(() => ({ test: 'event' }))
          ];
        }
      }

      const { app } = runWithTracking({ app: { Logic: Test } });

      expect(app.model).toMatchSnapshot();
    });
    it('should send update context from child', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return { ...evt } }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
        anotherTest: child(ContextChildLogic).create(b, c, a)
      });
      class ChildLogic {
        create() {
          return context(TEST_CONTEXT).get((ctx) => ({
            fromContext: ctx.test.ctx,
            argsFromContext: ctx.test.args
          }))
        }
        updateContext() {
          return context(TEST_CONTEXT).update(() => ({ test: 'event' }));
        }
      }
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { a: child(ChildLogic).create(1,2,3) },
          ];
        }
      }

      const { app } = runWithTracking({ app: { Logic: Test } });
      procOf(app.model.a).exec(logicOf(app.model.a).updateContext);

      expect(app.model).toMatchSnapshot();
    });
  });

  describe('#subscribe', () => {
    it('should not catch messages from context by default', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return evt }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
      });
      class ChildLogic {
        update = jest.fn()
      }
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { child: child(ChildLogic).create() },
            context(TEST_CONTEXT).update(() => ({ some: 'update' }))
          ];
        }
        update = jest.fn()
      }

      const { app, errors } = runWithTracking({ app: { Logic: Test } });
      const childLogic = logicOf(app.model.child);

      expect(app.model).toMatchSnapshot();
      expect(childLogic.update).toHaveBeenCalledTimes(0);
      expect(app.proc.logic.update).toHaveBeenCalledTimes(1);
    });

    it('should catch messages from context if subscribed', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return evt }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
      });
      class ChildLogic {
        create() {
          return context(TEST_CONTEXT).subscribe()
        }
        update = jest.fn()
      }
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { child: child(ChildLogic).create() },
            context(TEST_CONTEXT).update(() => ({ some: 'update' }))
          ];
        }
        update = jest.fn()
      }

      const { app, errors } = runWithTracking({ app: { Logic: Test } });
      const childLogic = logicOf(app.model.child);

      expect(app.model).toMatchSnapshot();
      expect(childLogic.update).toHaveBeenCalledTimes(1);
      expect(childLogic.update.mock.calls[0]).toMatchSnapshot();
      expect(app.proc.logic.update).toHaveBeenCalledTimes(1);
    });

    it('should remove subscription when logic destroyed', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return evt }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
      });
      class ChildLogic {
        create() { return context(TEST_CONTEXT).subscribe() }
        update = jest.fn()
      }
      class Test {
        create() {
          return [
            { ctx: context(TEST_CONTEXT).create(3,2,1) },
            { child: child(ChildLogic).create() }
          ];
        }
        update = jest.fn()
      }

      const { app, errors } = runWithTracking({ app: { Logic: Test } });
      const childLogic = logicOf(app.model.child);
      app.proc.exec({ child: null });
      app.proc.exec(context(TEST_CONTEXT).update(() => ({ some: 'update' })));

      expect(app.model).toMatchSnapshot();
      expect(childLogic.update).toHaveBeenCalledTimes(0);
      expect(procOf(app.model.ctx).handlers).toEqual([]);
      expect(app.proc.logic.update).toHaveBeenCalledTimes(1);
    });
    it('should throw an error if given context is not defined', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return evt }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
      });
      class ChildLogic {
        create() { return context(TEST_CONTEXT).subscribe() }
      }
      class Test {
        create() {
          return [
            { child: child(ChildLogic).create() }
          ];
        }
        update = jest.fn()
      }

      const { app, errors } = runWithTracking({ app: { Logic: Test }, expectErrors: true });

      expect(app.model).toMatchSnapshot();
      expect(errors).toMatchSnapshot();
    });
  });

  describe('#get', () => {
    it('should throw an error if given context is not defined', () => {
      class ContextChildLogic {
        create(...args) { return { ctx: 'yeah', args } }
        update(evt) { return evt }
      }
      const TEST_CONTEXT = (a, b, c) => ({
        test: child(ContextChildLogic).create(a, b, c),
      });
      class ChildLogic {
        create() { return context(TEST_CONTEXT).get(() => ({})) }
      }
      class Test {
        create() {
          return [
            { child: child(ChildLogic).create() }
          ];
        }
        update = jest.fn()
      }

      const { app, errors } = runWithTracking({ app: { Logic: Test }, expectErrors: true });

      expect(app.model).toMatchSnapshot();
      expect(errors).toMatchSnapshot();
    });
  });
});
