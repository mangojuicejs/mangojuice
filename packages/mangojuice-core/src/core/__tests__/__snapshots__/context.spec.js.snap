// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`child #create should be able to create context after child created 1`] = `
Object {
  "a": Object {
    "argsFromContext": Array [
      3,
      2,
      1,
    ],
    "fromContext": "yeah",
  },
  "ctx": Object {
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
    },
  },
}
`;

exports[`child #create should be able to create context after child created 2`] = `
Array [
  [Error: Context of given type does not exist],
]
`;

exports[`child #create should be able to create multiple contexts 1`] = `
Object {
  "anotherCtx": Object {
    "anotherTest": Object {
      "args": Array [
        2,
        1,
        3,
      ],
      "ctx": "yeah",
    },
  },
  "child": Object {
    "fromAnotherContext": Array [
      2,
      1,
      3,
    ],
    "fromTestContext": Array [
      3,
      2,
      1,
    ],
  },
  "ctx": Object {
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
    },
  },
}
`;

exports[`child #create should be able to create same context in different subtrees 1`] = `
Object {
  "a": Object {
    "ctx": Object {
      "anotherTest": Object {
        "args": Array [
          2,
          3,
          1,
        ],
        "ctx": "yeah",
      },
      "test": Object {
        "args": Array [
          1,
          2,
          3,
        ],
        "ctx": "yeah",
      },
    },
  },
  "b": Object {
    "ctx": Object {
      "anotherTest": Object {
        "args": Array [
          2,
          1,
          3,
        ],
        "ctx": "yeah",
      },
      "test": Object {
        "args": Array [
          3,
          2,
          1,
        ],
        "ctx": "yeah",
      },
    },
  },
}
`;

exports[`child #create should create context and available in child 1`] = `
Object {
  "a": Object {
    "argsFromContext": Array [
      3,
      2,
      1,
    ],
    "fromContext": "yeah",
  },
  "ctx": Object {
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
    },
  },
}
`;

exports[`child #create should destroy context with logic which created a context 1`] = `
Object {
  "ctx": Object {
    "anotherTest": Object {
      "args": Array [
        2,
        1,
        3,
      ],
      "ctx": "yeah",
    },
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
    },
  },
}
`;

exports[`child #create should destroy context with logic which created a context 2`] = `
Array [
  true,
  true,
  true,
  true,
]
`;

exports[`child #create should override a parent context if same created in child 1`] = `
Object {
  "anotherChild": Object {
    "fromContext": Array [
      3,
      2,
      1,
    ],
  },
  "child": Object {
    "child": Object {
      "fromContext": Array [
        2,
        2,
        2,
      ],
    },
    "ctx": Object {
      "anotherTest": Object {
        "args": Array [
          2,
          2,
          2,
        ],
        "ctx": "yeah",
      },
      "test": Object {
        "args": Array [
          2,
          2,
          2,
        ],
        "ctx": "yeah",
      },
    },
  },
  "ctx": Object {
    "anotherTest": Object {
      "args": Array [
        2,
        1,
        3,
      ],
      "ctx": "yeah",
    },
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
    },
  },
}
`;

exports[`child #create should override a parent context if same created in child 2`] = `Array []`;

exports[`child #create should throw an error if context already created - in root 1`] = `
Object {
  "ctx": Object {
    "anotherTest": Object {
      "args": Array [
        2,
        1,
        3,
      ],
      "ctx": "yeah",
    },
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
    },
  },
}
`;

exports[`child #create should throw an error if context already created - in root 2`] = `
Array [
  [Error: Context "TEST_CONTEXT" already created],
]
`;

exports[`child #get should throw an error if given context is not defined 1`] = `
Object {
  "child": Object {},
}
`;

exports[`child #get should throw an error if given context is not defined 2`] = `
Array [
  [Error: Context of given type does not exist],
]
`;

exports[`child #subscribe should catch messages from context if subscribed 1`] = `
Object {
  "child": Object {},
  "ctx": Object {
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
    },
  },
}
`;

exports[`child #subscribe should catch messages from context if subscribed 2`] = `
Array [
  Message {
    "some": "update",
  },
]
`;

exports[`child #subscribe should not catch messages from context by default 1`] = `
Object {
  "child": Object {},
  "ctx": Object {
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
    },
  },
}
`;

exports[`child #subscribe should remove subscription when logic destroyed 1`] = `
Object {
  "child": null,
  "ctx": Object {
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
    },
  },
}
`;

exports[`child #subscribe should throw an error if given context is not defined 1`] = `
Object {
  "child": Object {},
}
`;

exports[`child #subscribe should throw an error if given context is not defined 2`] = `
Array [
  [Error: Context of given type does not exist],
]
`;

exports[`child #update should send update context from child 1`] = `
Object {
  "a": Object {
    "argsFromContext": Array [
      3,
      2,
      1,
    ],
    "fromContext": "yeah",
  },
  "ctx": Object {
    "anotherTest": Object {
      "args": Array [
        2,
        1,
        3,
      ],
      "ctx": "yeah",
      "test": "event",
    },
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
      "test": "event",
    },
  },
}
`;

exports[`child #update should send update context from root 1`] = `
Object {
  "a": Object {
    "argsFromContext": Array [
      3,
      2,
      1,
    ],
    "fromContext": "yeah",
  },
  "ctx": Object {
    "anotherTest": Object {
      "args": Array [
        2,
        1,
        3,
      ],
      "ctx": "yeah",
      "test": "event",
    },
    "test": Object {
      "args": Array [
        3,
        2,
        1,
      ],
      "ctx": "yeah",
      "test": "event",
    },
  },
}
`;
